{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7403033e-9b40-453a-8d04-43744a332db5",
   "metadata": {},
   "source": [
    "Restriction site-Associated DNA sequencing (RADseq) is a high-throughput genotyping technique used in molecular biology and genomics. It is one method to generate reduced representation libraries, which are methods that allow us to prepare and sequence hundreds to thousands of genomic regions from across the genome without sequencing the entire genome. RADseq methods use restriction enzymes to cut up the genome and then sequence DNA regions that are adjacent to these cut sites. The idea is that within the same species or relatively closely related species, restriction enzyme cut sites should mostly be at the same places and allow for the selection of shared loci across samples without needing to develop sepecific probes.\n",
    "\n",
    "\n",
    "One of the most commonly used RADseq approaches, and the one that we'll use here, is double digest RADseq (ddRADseq) which uses two enzymes to cut up the genome and then a size selection step to further reduce the total set of total set of loci, which should ideally result in fewer loci that require less sequencing effort and that overlap among samples.\n",
    "\n",
    "we’ll be working with empirical double digest RADseq data [(Peterson et al. 2012)](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0037135) that I (Sean Harrington) generated as part of my PhD research at San Diego State University. The data are for a species of rattlesnake, the red diamond rattlesnake (Crotalus ruber), that is distributed across the Baja California peninsula and into southern California. I was interested in identifying if there is any population structure in *C. ruber* and inferring what population genetic and environmental forces have resulted in any existing structure. The data are single-end reads generated on an Illumina hiSeq. My analyses of these data are published in [Harrington et al. 2018](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0037135).\n",
    "\n",
    "\n",
    "\n",
    "The dataset is reasonably small and we should be able to quickly process and analyze it.\n",
    "\n",
    "We will use [ipyrad](https://ipyrad.readthedocs.io/en/master/) to process and assemble the raw data into alignments. ipyrad is a flexible python-based pipeline for taking various types of restriction-site associated data, processing them, and generated aligned datasets.\n",
    "\n",
    "iPyRad is capable of generating datasets either by mapping your raw reads to a reference genome or using a de novo assembly method that does not require a reference. We will use the de novo method today.\n",
    "\n",
    "If you need help with ipyrad outside of this workshop for specific issues, you can always post [here](https://app.gitter.im/#/room/#dereneaton_ipyrad:gitter.im). The developers are very responsive to queries.\n",
    "\n",
    "ipyrad is certainly not the only option for assembling RADseq data. [Stacks](https://catchenlab.life.illinois.edu/stacks/) and [dDocent](https://www.ddocent.com//) are other popular options, or there are various ways to manually assemble or map RADseq data.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "009069f0-bf36-4723-b83e-1f75699d6017",
   "metadata": {},
   "source": [
    "Files and basic setup\n",
    "\n",
    "The files we will use are:\n",
    "\n",
    "- all_ruber.fastq\n",
    "- barcodes_samples.txt\n",
    "- names_ruber_all.txt\n",
    "\n",
    "## Stuff about handling those files on the cloud"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3990df80-df0a-4e4a-a738-4e2ea53c68a6",
   "metadata": {},
   "source": [
    "## fastq format\n",
    "\n",
    "Before we start doing anything with the data, it's worth seeing what the raw data look like. The standard format for all raw data for genomic sequences is fastq.\n",
    "\n",
    "Let's take a look at the first 8 lines of the fastq file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0784010f-23fc-4c4a-ad8e-ef83163a90df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR6143937.sra.1 1 length=96\n",
      "TGATCGCTAANAGCAAATTGAGTCCCCTGCCCATCAGTTGATGATGTCATTGGTACTTTCTATTGTGTCAGGTCTTAACTTGCCATGTTTTTTTACTTTTATTA\n",
      "+SRR6143937.sra.1 1 length=96\n",
      "IIIIIIIIFD#24AFHJJJJJJIIJJJJJJJJJJJJJIJJJJJJJJJJJJJJJGHIJJJJJJJJJHIJJJJJ@FHIHIDHIHHHFHFFFFFFDDDDDCEDDDDA\n",
      "@SRR6143937.sra.2 2 length=96\n",
      "TGATCGCTTGNAGGGGGCGCATGAAGAGCGCAGGCACAGAGCAAGGCCCCGCCCTCCCCAGGGACTCATTGTGCAGTAACCGGATTGACTTCTCATGCACGCAG\n",
      "+SRR6143937.sra.2 2 length=96\n",
      "IIIIIIIIFF#22<DHIHJJJJJIJIJJJIJJJJHHHHFFFFFEECDDDDDDDDDDDDDDDDB<@BDDDEDEDDEDDD>CCDDBDDDDDDDDCDEEDDDDDDDB\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat ruber_data/all_ruber.fastq.gz | head -n 8"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d4b9cf6-c9e6-4128-8a1b-205cc9ee5113",
   "metadata": {},
   "source": [
    "- note that these reads are gzipped (end in .gz) you cannot directly look at them with head but instead need to use zcat, which reads gzipped files, and pipe the output to head. Fastqs are typically gzipped to save disk space and most genomics programs can read gzipped fastqs\n",
    "\n",
    "Each read from the sequencer is represented by 4 lines: the first 4 lines are the first read, the second set of 4 lines are the second read, etc. For each read, the first line is the header, and always starts with @. This contains a sequence identifier and various information about the read, often including information about the sequencing run. The second line, after the header, is the actual DNA sequence of the read. The next line always starts with + and may contain either no additional text, or the sequence identifier and extra information, as in the header. Line 4 for each read, following the + line, indicates the quality score for each DNA base in the read. This line will be exactly the same length as the DNA sequence in the second line, with e.g., the 4th character in this line corresponding to the quality of the 4th base in the sequence, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9428fbb7-8628-4fee-8434-41c2f949fc5f",
   "metadata": {},
   "source": [
    "## Running iPyRad\n",
    "\n",
    "\n",
    "First, we need to generate a params file that contains the parameters we need to specify for ipyrad. In your scripts directory, run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "36d27bc2-55db-474f-97e4-25782a6a836c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !ipyrad -n ruber_denovo  # commented because I've already edited, don't want to overwrite"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed742b9-61b6-44b1-b564-df0d16181796",
   "metadata": {},
   "source": [
    "This will create a params file with the defaults that ipyrad uses, we can modify these as we need . Whatever comes after the -n is what the assembly will be named\n",
    "\n",
    "Let’s go look at and edit that.\n",
    "\n",
    "\n",
    "### best way to edit a text file on gcp?\n",
    "## will need to edit paths in here\n",
    "\n",
    "We’ll change a few of these parameters:\n",
    "\n",
    "- [1]: This is where output will do, edit this to your ipyrad_out directory\n",
    "\n",
    "- [2]: this needs to reflect the path to the `all_ruber.fastq.gz` file\n",
    "\n",
    "- [3]: this needs to be the path to `barcodes_samples.txt`\n",
    "\n",
    "- [7]: dataype should be `ddrad`\n",
    "\n",
    "- [8]: restriction overhang is: `TGCAGG, GATC` these are the overhangs created by the restriction enzymes for ddRAD that was used for these data. I find these to be a pain to figure out, this is covered in the ipyrad params documentation\n",
    "\n",
    "- [14]: This is the clustering threshold for clustering reads into loci within samples. This is an important paramater that can have large effects on your final dataset. The default of `0.85` is good for phylogenetic datasets, but for population genetics, you will often want to use a higher threshold like 0.9 or 0.93. Let's use `0.9` here.\n",
    "\n",
    "- [27]: change to `*`, this will generate all output formats that ipyrad is currently capable of\n",
    "\n",
    "The rest of these are at generally reasonable values, although depending on your data, you may want to modify some of these. The parameters are all well documented [here](https://ipyrad.readthedocs.io/en/latest/6-params.html).\n",
    "\n",
    "For our final dataset, we'll want to set parameter [21] \"min_sample per locus\" to something higher to end up with a reasonable amount of missing data, but we'll deal with this later.\n",
    "\n",
    "We'll start by running steps 1-5:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c8507e3d-f742-4630-8014-c2938474b492",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " -------------------------------------------------------------\n",
      "  ipyrad [v.0.9.101]\n",
      "  Interactive assembly and analysis of RAD-seq data\n",
      " ------------------------------------------------------------- \n",
      "  Parallel connection | mbcpu-001.cluster: 16 cores\n",
      "  \n",
      "  Step 1: Demultiplexing fastq data to Samples\n",
      "  [####################] 100% 0:02:08 | sorting reads          \n",
      "  [####################] 100% 0:01:10 | writing/compressing    \n",
      "  \n",
      "  Step 2: Filtering and trimming reads\n",
      "  [####################] 100% 0:00:45 | processing reads     \n",
      "  \n",
      "  Step 3: Clustering/Mapping reads within samples\n",
      "  [####################] 100% 0:00:04 | dereplicating          \n",
      "  [####################] 100% 0:00:49 | clustering/mapping     \n",
      "  [####################] 100% 0:00:00 | building clusters      \n",
      "  [####################] 100% 0:00:00 | chunking clusters      \n",
      "  [####################] 100% 0:41:18 | aligning clusters      \n",
      "  [####################] 100% 0:00:04 | concat clusters        \n",
      "  [####################] 100% 0:00:00 | calc cluster stats     \n",
      "  \n",
      "  Step 4: Joint estimation of error rate and heterozygosity\n",
      "  [####################] 100% 0:00:19 | inferring [H, E]       \n",
      "  \n",
      "  Step 5: Consensus base/allele calling \n",
      "  Mean error  [0.00165 sd=0.00092]\n",
      "  Mean hetero [0.01052 sd=0.00152]\n",
      "  [####################] 100% 0:00:00 | calculating depths     \n",
      "  [####################] 100% 0:00:02 | chunking clusters      \n",
      "  [####################] 100% 0:00:52 | consens calling        \n",
      "  [####################] 100% 0:00:04 | indexing alleles       \n",
      "  Parallel connection closed.\n"
     ]
    }
   ],
   "source": [
    "# Run ipyrad with those parameters for steps 1-5 and using 16 cores\n",
    "!ipyrad -p params-ruber_denovo.txt -s 12345 -c 16\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b250661-b1d8-4e68-ae3e-c8da8fc4723e",
   "metadata": {},
   "source": [
    "This should take around 20 minutes. While that's running, familiarize yourself with the steps in ipyrad, which are thoroughly documented [here](https://ipyrad.readthedocs.io/en/master/7-outline.html).\n",
    "\n",
    "## No idea how long it'll actually take in the cloud\n",
    "~20 min as a slurm batch job with 8 cores, more like 40 min on our cluster via  jupyter notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a607b832-f2a6-4a78-9a60-655f137dba52",
   "metadata": {},
   "source": [
    "## Branching an assembly\n",
    "\n",
    "We only ran steps 1-5 above because the Fastq file that we started with includes mostly individuals of the red diamond rattlesnake, Crotalus ruber, but also a few outgroup taxa. Right now, we want to make a dataset that includes only C. ruber individuals that we can run some popgen analyses on in the next session.\n",
    "\n",
    "iPyRad includes functionality to make new “branches” of the assembly using different parameters and/or including/excluding different individuals, and we’ll take advantage of that functionality here.\n",
    "\n",
    "- If we wanted to include all samples in the same dataset, we could've just run all 7 steps at once.\n",
    "\n",
    "To create a new branch with only the desired individuals:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f143598-1bf3-402e-b637-87bc6a3ef0a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  loading Assembly: ruber_denovo\n",
      "  from saved path: /cluster/medbow/project/inbreh/radseq_cloud/ruber_denovo.json\n",
      "  creating a new branch called 'ruber_only_denovo' with 35 Samples\n",
      "  writing new params file to params-ruber_only_denovo.txt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# branch the assembly\n",
    "!ipyrad -p params-ruber_denovo.txt -b ruber_only_denovo ruber_data/names_ruber_all.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bbe72cd-1e3c-4a93-b6e7-8d6f2d49882f",
   "metadata": {},
   "source": [
    "This will use our old assembly and params file to generate a new branch, with params file `params-ruber_only_denovo.txt` that includes only samples in the `names_ruber_all.txt file`.\n",
    "\n",
    "We need to further edit this file to change parameter [21] “min_sample per locus”. The parameter defines how many how many individual samples a locus must have data for to include that locus in the final dataset. It controls the amount of missing data in the final dataset. Here, let's set this to `26` - this is about 75% of individuals and should result in a matrix that is ~75% or greater complete.\n",
    "\n",
    "Use your favorite text editor and make this change in the file params-ruber_only_denovo.txt:\n",
    "\n",
    "`26               ## [21] [min_samples_locus]: Min # samples per locus for output`\n",
    "\n",
    "Once that change has been made, run the final 2 steps in ipyrad. This should be very fast."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "556df84a-d2e6-4e40-9dae-64c81409535d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  loading Assembly: ruber_only_denovo\n",
      "  from saved path: /cluster/medbow/project/inbreh/radseq_cloud/ruber_only_denovo.json\n",
      "\n",
      " -------------------------------------------------------------\n",
      "  ipyrad [v.0.9.101]\n",
      "  Interactive assembly and analysis of RAD-seq data\n",
      " ------------------------------------------------------------- \n",
      "  Parallel connection | mbcpu-001.cluster: 8 cores\n",
      "  \n",
      "  Step 6: Clustering/Mapping across samples \n",
      "  [####################] 100% 0:00:04 | concatenating inputs   \n",
      "  [####################] 100% 0:00:12 | clustering across    \n",
      "  [####################] 100% 0:00:08 | building clusters      \n",
      "  [################### ]  96% 0:04:54 | aligning clusters    "
     ]
    }
   ],
   "source": [
    "!ipyrad -p params-ruber_only_denovo.txt -s 67 -c 16"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfdd6c2d-7431-47dd-b438-d6d3c834602a",
   "metadata": {},
   "source": [
    "## Examining the output\n",
    "\n",
    "\n",
    "Before you start analyzing your data, you should always take a look at the output stats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "82c41e82-c00e-4e33-886a-6d53d62bae12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "## The number of loci caught by each filter.\n",
      "## ipyrad API location: [assembly].stats_dfs.s7_filters\n",
      "\n",
      "                           total_filters applied_order retained_loci\n",
      "total_prefiltered_loci                 0             0         33335\n",
      "filtered_by_rm_duplicates            302           302         33033\n",
      "filtered_by_max_indels                 2             2         33031\n",
      "filtered_by_max_SNPs                   0             0         33031\n",
      "filtered_by_max_shared_het            17            17         33014\n",
      "filtered_by_min_sample             30516         30516          2498\n",
      "total_filtered_loci                30837         30837          2498\n",
      "\n",
      "\n",
      "## The number of loci recovered for each Sample.\n",
      "## ipyrad API location: [assembly].stats_dfs.s7_samples\n",
      "\n",
      "               sample_coverage\n",
      "SD_Field_0201             2463\n",
      "SD_Field_0255             1779\n",
      "SD_Field_0386             2253\n",
      "SD_Field_0491             2426\n",
      "SD_Field_0492             2432\n",
      "SD_Field_0493             2260\n",
      "SD_Field_0506               57\n",
      "SD_Field_0557             2479\n",
      "SD_Field_0598             2465\n",
      "SD_Field_0599             1820\n",
      "SD_Field_0642             2449\n",
      "SD_Field_0666             2116\n",
      "SD_Field_0983             2455\n",
      "SD_Field_1079             1604\n",
      "SD_Field_1205             2388\n",
      "SD_Field_1220             2476\n",
      "SD_Field_1225             2296\n",
      "SD_Field_1226             2474\n",
      "SD_Field_1381             2152\n",
      "SD_Field_1453             1081\n",
      "SD_Field_1878             2317\n",
      "SD_Field_1880             2465\n",
      "SD_Field_1899             2465\n",
      "SD_Field_1961             2216\n",
      "SD_Field_1988             2366\n",
      "SD_Field_1991             2300\n",
      "SD_Field_2127             2447\n",
      "SD_Field_2287             1796\n",
      "SD_Field_2383             2046\n",
      "SD_Field_2427             2469\n",
      "SD_Field_2789             2166\n",
      "SD_Field_2914             2320\n",
      "SD_Field_2968             2136\n",
      "SD_Field_3027             2476\n",
      "SD_Field_3124             2471\n",
      "\n",
      "\n",
      "## The number of loci for which N taxa have data.\n",
      "## ipyrad API location: [assembly].stats_dfs.s7_loci\n",
      "\n",
      "    locus_coverage  sum_coverage\n",
      "1                0             0\n",
      "2                0             0\n",
      "3                0             0\n",
      "4                0             0\n",
      "5                0             0\n",
      "6                0             0\n",
      "7                0             0\n",
      "8                0             0\n",
      "9                0             0\n",
      "10               0             0\n",
      "11               0             0\n",
      "12               0             0\n",
      "13               0             0\n",
      "14               0             0\n",
      "15               0             0\n",
      "16               0             0\n",
      "17               0             0\n",
      "18               0             0\n",
      "19               0             0\n",
      "20               0             0\n",
      "21               0             0\n",
      "22               0             0\n",
      "23               0             0\n",
      "24               0             0\n",
      "25               0             0\n",
      "26             284           284\n",
      "27             254           538\n",
      "28             219           757\n",
      "29             231           988\n",
      "30             187          1175\n",
      "31             181          1356\n",
      "32             212          1568\n",
      "33             353          1921\n",
      "34             541          2462\n",
      "35              36          2498\n",
      "\n",
      "\n",
      "The distribution of SNPs (var and pis) per locus.\n",
      "## var = Number of loci with n variable sites (pis + autapomorphies)\n",
      "## pis = Number of loci with n parsimony informative site (minor allele in >1 sample)\n",
      "## ipyrad API location: [assembly].stats_dfs.s7_snps\n",
      "## The \"reference\" sample is included if present unless 'exclude_reference=True'\n",
      "\n",
      "    var  sum_var   pis  sum_pis\n",
      "0   933        0  1357        0\n",
      "1   784      784   739      739\n",
      "2   438     1660   271     1281\n",
      "3   182     2206    83     1530\n",
      "4    90     2566    32     1658\n",
      "5    31     2721    13     1723\n",
      "6    21     2847     3     1741\n",
      "7    10     2917     0     1741\n",
      "8     5     2957     0     1741\n",
      "9     2     2975     0     1741\n",
      "10    1     2985     0     1741\n",
      "11    1     2996     0     1741\n",
      "\n",
      "\n",
      "## Final Sample stats summary\n",
      "               state  reads_raw  reads_passed_filter  clusters_total  clusters_hidepth  hetero_est  error_est  reads_consens  loci_in_assembly\n",
      "SD_Field_0201      7     432089               430182           45705             12432    0.009924   0.001556          11512              2463\n",
      "SD_Field_0255      7     140061               139403           29070              3558    0.011056   0.002011           3170              1779\n",
      "SD_Field_0386      7     369375               367788           27281              8115    0.006891   0.001189           7610              2253\n",
      "SD_Field_0491      7     432450               430646           57265             11375    0.011454   0.001629          10284              2426\n",
      "SD_Field_0492      7     431752               429453           56644             10908    0.011567   0.001629           9857              2432\n",
      "SD_Field_0493      7     455748               453957           41159             12341    0.009682   0.001321          11519              2260\n",
      "SD_Field_0506      7       8824                 8519            5710                98    0.012323   0.007015             78                57\n",
      "SD_Field_0557      7    1003459               999268           69485             26471    0.011766   0.001331          24584              2479\n",
      "SD_Field_0598      7     456199               454156           46949              9773    0.010340   0.001400           8890              2465\n",
      "SD_Field_0599      7     155064               154337           28866              3265    0.011766   0.001752           2902              1820\n",
      "SD_Field_0642      7     390743               389052           44665              9138    0.009436   0.001418           8358              2449\n",
      "SD_Field_0666      7     210617               209592           26775              5116    0.010158   0.001560           4630              2116\n",
      "SD_Field_0983      7     475105               472944           48252             11342    0.009387   0.001346          10446              2455\n",
      "SD_Field_1079      7      88291                87831           14633              2706    0.009807   0.002314           2458              1604\n",
      "SD_Field_1205      7     376651               374972           48182              9623    0.010468   0.001464           8689              2388\n",
      "SD_Field_1220      7    1565453              1558432           55366             28703    0.010125   0.001073          26898              2476\n",
      "SD_Field_1225      7     552035               549517           46068             13669    0.011766   0.001385          12668              2296\n",
      "SD_Field_1226      7     696782               693836           45775             16149    0.011396   0.001359          15034              2474\n",
      "SD_Field_1381      7     287449               285831           33245              6466    0.008900   0.001542           5905              2152\n",
      "SD_Field_1453      7     176716               175874           11403              2366    0.008582   0.001090           2186              1081\n",
      "SD_Field_1878      7     384964               383414           32659              8420    0.007782   0.001153           7831              2317\n",
      "SD_Field_1880      7     598416               595794           60189             16264    0.012544   0.001581          15022              2465\n",
      "SD_Field_1899      7     517578               515278           40206             12243    0.008690   0.001248          11497              2465\n",
      "SD_Field_1961      7     397383               395521           47965             10579    0.010093   0.001009           9695              2216\n",
      "SD_Field_1988      7     832987               829093           62008             21870    0.010406   0.001546          20474              2366\n",
      "SD_Field_1991      7     614203               611640           45208             15863    0.010965   0.001345          14830              2300\n",
      "SD_Field_2127      7     518329               515765           61341             16185    0.012425   0.001986          14887              2447\n",
      "SD_Field_2287      7     165292               164632           22571              3892    0.009874   0.001543           3543              1796\n",
      "SD_Field_2383      7     138325               137437           18531              4491    0.008871   0.001912           4096              2046\n",
      "SD_Field_2427      7     483472               481385           46815             11510    0.009903   0.001391          10611              2469\n",
      "SD_Field_2789      7     372514               370928           38156              9763    0.008194   0.001333           9022              2166\n",
      "SD_Field_2914      7     251560               250354           31273              6023    0.010209   0.001585           5429              2320\n",
      "SD_Field_2968      7     238885               237497           46411              5994    0.011158   0.002240           5291              2136\n",
      "SD_Field_3027      7     593722               591229           47416             12909    0.010574   0.001202          11935              2476\n",
      "SD_Field_3124      7     803222               799999           47329             18216    0.011307   0.001401          16866              2471\n",
      "\n",
      "\n",
      "## Alignment matrix statistics:\n",
      "snps matrix size: (35, 2996), 14.25% missing sites.\n",
      "sequence matrix size: (35, 224550), 12.95% missing sites.\n"
     ]
    }
   ],
   "source": [
    "## JUST LOOK AT FILE IN GCP?????????\n",
    "\n",
    "!cat ruber_only_denovo_outfiles/ruber_only_denovo_stats.txt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fa34a23-89ce-4d02-b40b-29d732ffdc1f",
   "metadata": {},
   "source": [
    "There should be about 2498 loci recovered in the assembly (last column of row `total_filtered_loci`). If we scroll down a bit in the table `The number of loci recovered for each Sample`, we can see that SD_Field_0506 has almost no loci shared with other samples, and SD_Field_1453 has only about half as many loci as most samples. We’ll want to remove these samples before moving on. Note that SD_Field_0506 is an obviously failed sample, but for SD_Field_1453, you would likely want to try out some preliminary downstream analyses with and without this sample – I’ve already analyzed these data and decided it’s best to remove it."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d973e6fa-16da-4f6b-9cc5-122ca373fdee",
   "metadata": {},
   "source": [
    "## Branch to remove low data samples\n",
    "\n",
    "Start by making a new names file to exclude SD_Field_0506 and SD_Field_1453 called `names_ruber_reduced.txt` and delete the lines containing `SD_Field_0506` and `SD_Field_1453`.\n",
    "\n",
    "### best way to edit that file in gcp????\n",
    "\n",
    "Then do the branching and run step 7 on that new branch:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ef23bd8-7fba-4ede-938b-2d9e373a92a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  loading Assembly: ruber_only_denovo\n",
      "  from saved path: /cluster/medbow/project/inbreh/radseq_cloud/ruber_only_denovo.json\n",
      "  creating a new branch called 'ruber_reduced_denovo' with 33 Samples\n",
      "  writing new params file to params-ruber_reduced_denovo.txt\n",
      "\n",
      "  loading Assembly: ruber_reduced_denovo\n",
      "  from saved path: /cluster/medbow/project/inbreh/radseq_cloud/ruber_reduced_denovo.json\n",
      "\n",
      " -------------------------------------------------------------\n",
      "  ipyrad [v.0.9.101]\n",
      "  Interactive assembly and analysis of RAD-seq data\n",
      " ------------------------------------------------------------- \n",
      "  Parallel connection | mbcpu-001.cluster: 8 cores\n",
      "  \n",
      "  Step 7: Filtering and formatting output files \n",
      "  [####################] 100% 0:00:36 | applying filters       \n",
      "  [####################] 100% 0:00:03 | building arrays        \n",
      "  [####################] 100% 0:00:04 | writing conversions    \n",
      "  [####################] 100% 0:00:32 | indexing vcf depths    \n",
      "  [####################] 100% 0:00:06 | writing vcf output     \n",
      "  Parallel connection closed.\n"
     ]
    }
   ],
   "source": [
    "!ipyrad -p params-ruber_only_denovo.txt -b ruber_reduced_denovo ruber_data/names_ruber_reduced.txt\n",
    "\n",
    "!ipyrad -p params-ruber_reduced_denovo.txt -s 7 -c 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "84ec68d9-3b1f-4070-b9a8-57e2ae6b2620",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "## The number of loci caught by each filter.\n",
      "## ipyrad API location: [assembly].stats_dfs.s7_filters\n",
      "\n",
      "                           total_filters applied_order retained_loci\n",
      "total_prefiltered_loci                 0             0         33335\n",
      "filtered_by_rm_duplicates            301           301         33034\n",
      "filtered_by_max_indels                 1             1         33033\n",
      "filtered_by_max_SNPs                   0             0         33033\n",
      "filtered_by_max_shared_het            19            19         33014\n",
      "filtered_by_min_sample             30563         30563          2451\n",
      "total_filtered_loci                30884         30884          2451\n",
      "\n",
      "\n",
      "## The number of loci recovered for each Sample.\n",
      "## ipyrad API location: [assembly].stats_dfs.s7_samples\n",
      "\n",
      "               sample_coverage\n",
      "SD_Field_0201             2416\n",
      "SD_Field_0255             1775\n",
      "SD_Field_0386             2213\n",
      "SD_Field_0491             2383\n",
      "SD_Field_0492             2389\n",
      "SD_Field_0493             2245\n",
      "SD_Field_0557             2434\n",
      "SD_Field_0598             2420\n",
      "SD_Field_0599             1786\n",
      "SD_Field_0642             2406\n",
      "SD_Field_0666             2079\n",
      "SD_Field_0983             2412\n",
      "SD_Field_1079             1571\n",
      "SD_Field_1205             2346\n",
      "SD_Field_1220             2432\n",
      "SD_Field_1225             2277\n",
      "SD_Field_1226             2428\n",
      "SD_Field_1381             2109\n",
      "SD_Field_1878             2273\n",
      "SD_Field_1880             2421\n",
      "SD_Field_1899             2420\n",
      "SD_Field_1961             2207\n",
      "SD_Field_1988             2347\n",
      "SD_Field_1991             2284\n",
      "SD_Field_2127             2404\n",
      "SD_Field_2287             1759\n",
      "SD_Field_2383             2007\n",
      "SD_Field_2427             2423\n",
      "SD_Field_2789             2154\n",
      "SD_Field_2914             2291\n",
      "SD_Field_2968             2098\n",
      "SD_Field_3027             2429\n",
      "SD_Field_3124             2424\n",
      "\n",
      "\n",
      "## The number of loci for which N taxa have data.\n",
      "## ipyrad API location: [assembly].stats_dfs.s7_loci\n",
      "\n",
      "    locus_coverage  sum_coverage\n",
      "1                0             0\n",
      "2                0             0\n",
      "3                0             0\n",
      "4                0             0\n",
      "5                0             0\n",
      "6                0             0\n",
      "7                0             0\n",
      "8                0             0\n",
      "9                0             0\n",
      "10               0             0\n",
      "11               0             0\n",
      "12               0             0\n",
      "13               0             0\n",
      "14               0             0\n",
      "15               0             0\n",
      "16               0             0\n",
      "17               0             0\n",
      "18               0             0\n",
      "19               0             0\n",
      "20               0             0\n",
      "21               0             0\n",
      "22               0             0\n",
      "23               0             0\n",
      "24               0             0\n",
      "25               0             0\n",
      "26             300           300\n",
      "27             255           555\n",
      "28             215           770\n",
      "29             222           992\n",
      "30             188          1180\n",
      "31             203          1383\n",
      "32             258          1641\n",
      "33             810          2451\n",
      "\n",
      "\n",
      "The distribution of SNPs (var and pis) per locus.\n",
      "## var = Number of loci with n variable sites (pis + autapomorphies)\n",
      "## pis = Number of loci with n parsimony informative site (minor allele in >1 sample)\n",
      "## ipyrad API location: [assembly].stats_dfs.s7_snps\n",
      "## The \"reference\" sample is included if present unless 'exclude_reference=True'\n",
      "\n",
      "   var  sum_var   pis  sum_pis\n",
      "0  921        0  1330        0\n",
      "1  778      778   726      726\n",
      "2  425     1628   270     1266\n",
      "3  176     2156    82     1512\n",
      "4   85     2496    30     1632\n",
      "5   29     2641    10     1682\n",
      "6   22     2773     3     1700\n",
      "7   10     2843     0     1700\n",
      "8    3     2867     0     1700\n",
      "9    2     2885     0     1700\n",
      "\n",
      "\n",
      "## Final Sample stats summary\n",
      "               state  reads_raw  reads_passed_filter  clusters_total  clusters_hidepth  hetero_est  error_est  reads_consens  loci_in_assembly\n",
      "SD_Field_0201      7     432089               430182           45705             12432    0.009924   0.001556          11512              2416\n",
      "SD_Field_0255      7     140061               139403           29070              3558    0.011056   0.002011           3170              1775\n",
      "SD_Field_0386      7     369375               367788           27281              8115    0.006891   0.001189           7610              2213\n",
      "SD_Field_0491      7     432450               430646           57265             11375    0.011454   0.001629          10284              2383\n",
      "SD_Field_0492      7     431752               429453           56644             10908    0.011567   0.001629           9857              2389\n",
      "SD_Field_0493      7     455748               453957           41159             12341    0.009682   0.001321          11519              2245\n",
      "SD_Field_0557      7    1003459               999268           69485             26471    0.011766   0.001331          24584              2434\n",
      "SD_Field_0598      7     456199               454156           46949              9773    0.010340   0.001400           8890              2420\n",
      "SD_Field_0599      7     155064               154337           28866              3265    0.011766   0.001752           2902              1786\n",
      "SD_Field_0642      7     390743               389052           44665              9138    0.009436   0.001418           8358              2406\n",
      "SD_Field_0666      7     210617               209592           26775              5116    0.010158   0.001560           4630              2079\n",
      "SD_Field_0983      7     475105               472944           48252             11342    0.009387   0.001346          10446              2412\n",
      "SD_Field_1079      7      88291                87831           14633              2706    0.009807   0.002314           2458              1571\n",
      "SD_Field_1205      7     376651               374972           48182              9623    0.010468   0.001464           8689              2346\n",
      "SD_Field_1220      7    1565453              1558432           55366             28703    0.010125   0.001073          26898              2432\n",
      "SD_Field_1225      7     552035               549517           46068             13669    0.011766   0.001385          12668              2277\n",
      "SD_Field_1226      7     696782               693836           45775             16149    0.011396   0.001359          15034              2428\n",
      "SD_Field_1381      7     287449               285831           33245              6466    0.008900   0.001542           5905              2109\n",
      "SD_Field_1878      7     384964               383414           32659              8420    0.007782   0.001153           7831              2273\n",
      "SD_Field_1880      7     598416               595794           60189             16264    0.012544   0.001581          15022              2421\n",
      "SD_Field_1899      7     517578               515278           40206             12243    0.008690   0.001248          11497              2420\n",
      "SD_Field_1961      7     397383               395521           47965             10579    0.010093   0.001009           9695              2207\n",
      "SD_Field_1988      7     832987               829093           62008             21870    0.010406   0.001546          20474              2347\n",
      "SD_Field_1991      7     614203               611640           45208             15863    0.010965   0.001345          14830              2284\n",
      "SD_Field_2127      7     518329               515765           61341             16185    0.012425   0.001986          14887              2404\n",
      "SD_Field_2287      7     165292               164632           22571              3892    0.009874   0.001543           3543              1759\n",
      "SD_Field_2383      7     138325               137437           18531              4491    0.008871   0.001912           4096              2007\n",
      "SD_Field_2427      7     483472               481385           46815             11510    0.009903   0.001391          10611              2423\n",
      "SD_Field_2789      7     372514               370928           38156              9763    0.008194   0.001333           9022              2154\n",
      "SD_Field_2914      7     251560               250354           31273              6023    0.010209   0.001585           5429              2291\n",
      "SD_Field_2968      7     238885               237497           46411              5994    0.011158   0.002240           5291              2098\n",
      "SD_Field_3027      7     593722               591229           47416             12909    0.010574   0.001202          11935              2429\n",
      "SD_Field_3124      7     803222               799999           47329             18216    0.011307   0.001401          16866              2424\n",
      "\n",
      "\n",
      "## Alignment matrix statistics:\n",
      "snps matrix size: (33, 2885), 10.04% missing sites.\n",
      "sequence matrix size: (33, 220324), 8.74% missing sites.\n"
     ]
    }
   ],
   "source": [
    "# Look at the stats for the new assembly:\n",
    "!cat ruber_reduced_denovo_outfiles/ruber_reduced_denovo_stats.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a054a01-aa43-458f-a8c8-d3bd7a8f8170",
   "metadata": {},
   "source": [
    "You should now see a slight decrease in the number of loci (I see 2451), but pretty good coverage across individuals, with no single sample having maassive amounts of missing data. This looks like a good dataset to move forward with.\n",
    "\n",
    "We have all sorts of variously formatted data files in the output directory."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ipyrad-env",
   "language": "python",
   "name": "ipyrad-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
